pkg/objcache/objcache.go

```diff
@@ -59,20 +78,40 @@ type Cache struct {
 	entries map[string]buffer
 
 	// Expiration in time duration.
-	expiry time.Duration
+	expiration time.Duration
+
+	// Cleanup interval in time duration.
+	cleanupInterval time.Duration
+
+	// Janitor maintains carries cleanup interval
+	// and stop channel for cleaning up janitor routine.
+	janitor *janitor
 }
 
-// New creates an inmemory cache
-//
-// maxSize is used for expiring objects before we run out of memory
-// expiration is used for expiration of a key from cache
-func New(maxSize uint64, expiry time.Duration) *Cache {
-	return &Cache{
-		mutex:   &sync.RWMutex{},
-		maxSize: maxSize,
-		entries: make(map[string]buffer),
-		expiry:  expiry,
+// New - Return a new cache with a given default expiration duration and cleanup
+// interval. If the expiration duration is less than one (or NoExpiration),
+// the items in the cache never expire (by default), and must be deleted
+// manually. If the cleanup interval is less than one, expired items are not
+// deleted from the cache before calling c.DeleteExpired().
+func New(maxSize uint64, expiration time.Duration, cleanupInterval time.Duration) *Cache {
+	c := &cache{
+		mutex:           &sync.RWMutex{},
+		maxSize:         maxSize,
+		entries:         make(map[string]buffer),
+		expiration:      expiration,
+		cleanupInterval: cleanupInterval,
+	}
+	// This trick ensures that the janitor goroutine (which--granted it
+	// was enabled--is running DeleteExpired on c forever) does not keep
+	// the returned C object from being garbage collected. When it is
+	// garbage collected, the finalizer stops the janitor goroutine, after
+	// which c can be collected.
+	C := &Cache{c}
```

Comment:
This trick is hard to read. You can simply have `func (c *Cache) Destroy()` that stops the janitor goroutine.